File: agent-driven-architecture-main/examples/kotlin/block/artifact/src/main/kotlin/adr/blocks/artifact/Project.kt
// ── blocks/artifact/project — TWO pure projections of the SAME slice ───────
// Note what the CONTEXT projection contributes: a COUNT, never the lines (§5.2).
// That is the whole reason a long session cannot inflate the reasoner's input by
// writing more findings.
//
// THIS BLOCK HAS NO `contextLines` AT ALL, and that measured fact is why the shared
// `Block` interface declares only `arm` and `view`. A four-role interface would have
// forced this block to fake a role it does not have, or forced the root to special-case
// it — the exact per-block special-casing the interface exists to prevent.

package adr.blocks.artifact

data class ArtifactView(
    val lines: List<String>,
    val state: String,
    val canSeal: Boolean,
    val sealed: Boolean,
)

internal class ArtifactProjection {

    fun view(slice: ArtifactSlice): ArtifactView = when (val seal = slice.seal) {
        SealStatus.Draft -> ArtifactView(
            lines = slice.lines.map { it.text },
            state = "draft",
            canSeal = true,
            sealed = false,
        )

        is SealStatus.Sealing -> ArtifactView(
            lines = slice.lines.map { it.text },
            state = "seal requested by ${seal.requestedBy.id}",
            canSeal = false,
            sealed = false,
        )

        is SealStatus.Sealed -> ArtifactView(
            lines = slice.lines.map { it.text },
            state = "sealed at ${seal.at.value} by ${seal.by.id}",
            canSeal = false,
            sealed = true,
        )
    }

    /** The artifact enters the reasoner's Context by COUNT only — never by content. */
    fun lineCount(slice: ArtifactSlice): Int = slice.lines.size
}

File: agent-driven-architecture-main/examples/typescript/src/blocks/artifact/project.ts
// ── blocks/artifact/project — the two pure projections ─────────────────────
// The context projection contributes a COUNT and a status, never the lines
// themselves (§5.2's growth bound): the reasoner's input must not grow with the
// artifact.

import { bounded, MAX_CONTEXT_LINES_PER_BLOCK } from "@adr/spine/pure/context";
import type { ArtifactLine, ArtifactSlice, SealStatus } from "./slice";

export interface ArtifactView {
  readonly lines: readonly ArtifactLine[];
  readonly lineCount: number;
  readonly seal: string;
  readonly canSeal: boolean;
}

export function artifactView(slice: ArtifactSlice): ArtifactView {
  return {
    lines: slice.lines,
    lineCount: slice.lines.length,
    seal: sealLabel(slice.seal),
    // the PARENT-declared property again, not an `===` against one variant
    canSeal: slice.seal.requestedBy !== null,
  };
}

function sealLabel(seal: SealStatus): string {
  switch (seal.kind) {
    case "Draft":
      return "draft";
    case "Sealing":
      return "seal requested";
    case "Sealed":
      return `sealed at ${seal.at}`;
    default: {
      const _never: never = seal;
      return _never;
    }
  }
}

export function artifactContextLines(
  slice: ArtifactSlice,
  max: number = MAX_CONTEXT_LINES_PER_BLOCK,
): readonly string[] {
  return bounded([`work product: ${slice.lines.length} line(s), ${sealLabel(slice.seal)}`], max);
}
